### PR TITLE
[v0] dotnet library for starknet

### PR DESCRIPTION
**Pull request by trouaux**

### CODE REVIEW

File: src/StarkNet.Core/HttpProvider.cs [hunk 1]
```diff
@@ -1,12 +1,214 @@
...
+        feederGatewayUrl = _feederGatewayUrl;
+        gatewayUrl = _gatewayUrl;
+        httpClient = _httpClient;
+    }
+
+    #region PUBLIC
+    public async Task<bool> IsAlive()
+    {
+        const string path = "/is_alive";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 2]
```diff
@@ -1,12 +1,214 @@
...
+
+        HttpResponseMessage response = await httpClient.GetAsync(requestUri);
+
+        return response.IsSuccessStatusCode;
+    }
+
+    public async Task<ContractAddresses> GetContractAddresses()
+    {
+        const string path = "/get_contract_addresses";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 3]
```diff
@@ -1,12 +1,214 @@
...
+        
+        return (ContractAddresses)contractAddresses;
+    }
+
+    public async Task<Block> GetBlock(string? blockIdentifier = null)
+    {
+        const string path = "/get_block?";
+        Uri requestUri;
+
+        if (blockIdentifier != null)
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Use `!String.IsNullOrWhiteSpace`

File: src/StarkNet.Core/HttpProvider.cs [hunk 4]
```diff
@@ -1,12 +1,214 @@
...
+
+        var code = await response.Content.ReadAsAsync(typeof(Code));
+
+        return (Code)code;
+    }
+
+    public async Task<string> GetStorageAt(string contractAddress, string key, string blockIdentifier)
+    {
+        const string path = "/get_storage_at?";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 5]
```diff
@@ -1,12 +1,214 @@
...
+
+        var code = await response.Content.ReadAsAsync(typeof(string));
+
+        return (string)code;
+    }
+
+    public async Task<TransactionStatus> GetTransactionStatus(string txHash)
+    {
+        const string path = "/get_transaction_status?";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 6]
```diff
@@ -1,12 +1,214 @@
...
+        var code = await response.Content.ReadAsAsync(typeof(TransactionStatus));
+
+        return (TransactionStatus)code;
+    }
+
+    //txId compatibility
+    public async Task<TransactionReceipt> GetTransactionReceipt(string txHash, string txId)
+    {
+        const string path = "/get_transaction_receipt?";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 7]
```diff
@@ -1,12 +1,214 @@
...
+        var code = await response.Content.ReadAsAsync(typeof(TransactionReceipt));
+
+        return (TransactionReceipt)code;
+    }
+
+    //txid ?
+    public async Task<Transaction> GetTransaction(string txHash)
+    {
+        const string path = "/get_transaction?";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 8]
```diff
@@ -1,12 +1,214 @@
...
+        return (TransactionStatus)code;
+    }
+
+    //txId compatibility
+    public async Task<TransactionReceipt> GetTransactionReceipt(string txHash, string txId)
+    {
+        const string path = "/get_transaction_receipt?";
+        Uri requestUri;
+
+        requestUri = new Uri(baseUrl + feederGatewayUrl + path + "transactionHash=" + txHash);
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Should check if txHash is null or whitespace and maybe throw if it is?

File: src/StarkNet.Core/HttpProvider.cs [hunk 9]
```diff
@@ -1,12 +1,214 @@
...
+        return (TransactionReceipt)code;
+    }
+
+    //txid ?
+    public async Task<Transaction> GetTransaction(string txHash)
+    {
+        const string path = "/get_transaction?";
+        Uri requestUri;
+
+        requestUri = new Uri(baseUrl + feederGatewayUrl + path + "transactionHash=" + txHash);
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Should check if txHash is null or whitespace and maybe throw if it is?

File: src/StarkNet.Core/HttpProvider.cs [hunk 10]
```diff
@@ -1,12 +1,214 @@
...
+        return (Code)code;
+    }
+
+    public async Task<string> GetStorageAt(string contractAddress, string key, string blockIdentifier)
+    {
+        const string path = "/get_storage_at?";
+        Uri requestUri;
+        
+        var identifier = formatBlockIdentifier(blockIdentifier);
+        requestUri = new Uri(baseUrl + feederGatewayUrl + path + "contractAddress=" + contractAddress + "&key=" + key + identifier);
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Should check if key is null or whitespace and maybe throw if it is?

File: src/StarkNet.Core/HttpProvider.cs [hunk 11]
```diff
@@ -1,12 +1,214 @@
...
+        return (Transaction)code;
+    }
+
+    //txid ?
+    public async Task<TransactionTrace> GetTransactionTrace(string txHash)
+    {
+        const string path = "/get_transaction_trace?";
+        Uri requestUri;
+
+        requestUri = new Uri(baseUrl + feederGatewayUrl + path + "transactionHash=" + txHash);
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Should check if txHash is null or whitespace and maybe throw if it is?

File: src/StarkNet.Core/HttpProvider.cs [hunk 12]
```diff
@@ -1,12 +1,214 @@
...
+        var code = await response.Content.ReadAsAsync(typeof(Transaction));
+
+        return (Transaction)code;
+    }
+
+    //txid ?
+    public async Task<TransactionTrace> GetTransactionTrace(string txHash)
+    {
+        const string path = "/get_transaction_trace?";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 13]
```diff
@@ -1,12 +1,214 @@
...
+        var code = await response.Content.ReadAsAsync(typeof(TransactionTrace));
+
+        return (TransactionTrace)code;
+    }
+
+    public async Task<string> AddTransaction(string body)
     {
-        public void Dispose()
+        const string path = "/add_transaction";
+        Uri requestUri;
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs]: Declare and assign variable on the same line

File: src/StarkNet.Core/HttpProvider.cs [hunk 14]
```diff
@@ -1,12 +1,214 @@
...
+
+        requestUri = new Uri(baseUrl + gatewayUrl + path);
+
+        var stringContent = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs:173]: Should check if body is null or whitespace and throw?

File: src/StarkNet.Core/HttpProvider.cs [hunk 15]
```diff
@@ -1,12 +1,214 @@
...
+
+        return (string)code;
+    }
+
+    #endregion
+
+    #region PRIVATE
+    private string formatBlockIdentifier(string blockIdentifier)
+    {
+        if (blockIdentifier.StartsWith("0x"))
```
Comment (JuanuMusic) [inline src/StarkNet.Core/HttpProvider.cs:188]: Should check if blockIdentifier is null or whitespace?

File: tests/StarkNet.Core.LocalTests/Program.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+        //var block = await provider.GetBlock("0x2b9d0650d258656864f887eef9a28b9d318c1e5115782a2c7cc940333ee4acc").ConfigureAwait(false);
+        //var code = await provider.GetCode("0x01d1f307c073bb786a66e6e042ec2a9bdc385a3373bb3738d95b966d5ce56166", "").ConfigureAwait(false);
+        //var storage = await provider.GetStorageAt("0x01d1f307c073bb786a66e6e042ec2a9bdc385a3373bb3738d95b966d5ce56166", "0", "36663").ConfigureAwait(false);
+        //var transactionTrace = await provider.GetTransactionTrace("0x37013e1cb9c133e6fe51b4b371b76b317a480f56d80576730754c1662582348").ConfigureAwait(false);
+        //var transaction = await provider.GetTransaction("0x37013e1cb9c133e6fe51b4b371b76b317a480f56d80576730754c1662582348");
+        //var transactionReceipt = await provider.GetTransactionReceipt("0x37013e1cb9c133e6fe51b4b371b76b317a480f56d80576730754c1662582348", "");
+        //var transactionStatus = await provider.GetTransactionStatus("0x37013e1cb9c133e6fe51b4b371b76b317a480f56d80576730754c1662582348");
+
+        string text = System.IO.File.ReadAllText(@"C:\Users\tr\Documents\repos\starknet-dotnet\tests\StarkNet.Core.LocalTests\json1.json");
```
Comment (JuanuMusic) [inline tests/StarkNet.Core.LocalTests/Program.cs]: Can this path be turned into a relative path?

Comment (trouaux) [reply tests/StarkNet.Core.LocalTests/Program.cs]: this is really just some personal local test, we should build proper integration tests imo

Comment (franceccon) [reply tests/StarkNet.Core.LocalTests/Program.cs]: I think it's enough to use relative path and then change the project to copy the files on build.

**Comment by fracek** [comment]
Looks good! I think the field in the struct should be `CamelCase` and not `snake_case`. I also believe we should use the more modern `System.Text.Json` for json. The good news is that you can solve both issues in one! 

Use [`HttpContentJsonExtension`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.json.httpcontentjsonextensions.readfromjsonasync?view=net-6.0) to read the response into a specific type (without the need to type cast)

For example:
 ```cs
var block = await response.Content.ReadAsAsync(typeof(Block));
return (Block)block;
```
becomes 

```cs
return await response.Content.ReadFromJsonAsync<Block>();
```

The last step is to change the model types to use CamelCase for fields and snake case for serialization and deserialization

```cs
public class BlockTransaction
{
    public string contract_address { get; set; }
    public string entry_point_selector { get; set; }
    public string entry_point_type { get; set; }
    public List<string> calldata { get; set; }
} 
```

becomes

```cs
public class BlockTransaction
{
    [JsonPropertyName("contract_address")]
    public string ContractAddress { get; set; }

    [JsonPropertyName("entry_point_selector")]
    public string EntryPointSelector { get; set; }

    [JsonPropertyName("entry_point_type")]
    public string EntryPointType { get; set; }

    [JsonPropertyName("calldata")]
    public List<string> Calldata { get; set; }
} 
```

**Comment by fracek** [comment]
Also thank you @JuanuMusic for the help with the PR!

**Comment by trouaux** [comment]
@fracek followed your suggestions to read JSON directly, looks better now 

@JuanuMusic I think I went through most of your comments

definitely see some stuff to improve still, but maybe worth doing this in new smaller PRs, need to think about that

File: src/StarkNet.Core/HttpProvider.cs [hunk 17]
```diff
@@ -1,12 +1,198 @@
...
+    {
+        const string path = "/get_contract_addresses";
+
+        Uri requestUri = new Uri(baseUrl + feederGatewayUrl + path);
+
+        HttpResponseMessage response = await httpClient.GetAsync(requestUri);
+
+        var contractAddresses = await response.Content.ReadAsAsync(typeof(ContractAddresses));
+        
+        return (ContractAddresses)contractAddresses;
```
Comment (franceccon) [inline src/StarkNet.Core/HttpProvider.cs:62]: Missed this one.

**Review by fracek — APPROVED** [review]

**Comment by fracek** [comment]
I will merge so that everyone can start working on their piece. Thank you for kickstarting the project!

**Comment by fracek** [comment]
I merged this PR manually because there were some tiny issues.

**Pull request closed without merging**